components/percona_telemetry/storage.cc

```diff
@@ -0,0 +1,135 @@
+#include <fstream>
+
+#include "logger.h"
+#include "storage.h"
+
+#include <sys/stat.h>
+#include <ctime>
+#include <filesystem>
+#include <vector>
+
+#include <boost/uuid/random_generator.hpp>
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_io.hpp>
+
+namespace fs = std::filesystem;
+
+static std::string TMP_FILE_EXTENSION(".tmp");
+static std::string JSON_FILE_EXTENSION(".json");
+
+class file_stream_exception : public std::exception {
+ public:
+  file_stream_exception(const char *what) : what_(what) {}
+  file_stream_exception(const std::string &what) : what_(what) {}
+  const char *what() const noexcept override { return what_.c_str(); }
+
+ private:
+  std::string what_;
+};
+
+static std::string random_uuid() {
+  static thread_local boost::uuids::random_generator gen;
+  auto generated = gen();
+  return boost::uuids::to_string(generated);
+}
+
+Storage::Storage(Config &config, Logger &logger)
+    : config_(config), logger_(logger), uuid_(random_uuid()) {}
+
+void Storage::clean_old_reports(long current_time) {
+  long oldest_timestamp = current_time - config_.history_keep_interval();
+  long oldest_unconditional_cleanup_timestamp =
+      current_time - config_.unconditional_history_cleanup_interval();
+
+  std::vector<std::string> files_to_delete;
+
+  for (const auto &entry :
+       fs::directory_iterator(config_.telemetry_storage_dir_path())) {
+    if (entry.path().extension() == TMP_FILE_EXTENSION ||
+        entry.path().extension() == JSON_FILE_EXTENSION) {
+      auto filename = entry.path().stem().string();
+      // filename should be <timestamp>-<uuid_>.json/tmp
+      size_t delimiter_pos = filename.find("-");
+      if (delimiter_pos == std::string::npos || delimiter_pos == 0 ||
+          delimiter_pos == filename.length() - 1) {
+        logger_.warning("Skipping file deletion %s", filename.c_str());
+        continue;
+      }
+      auto file_timestamp_str = filename.substr(0, delimiter_pos);
+      auto file_uuid = filename.substr(delimiter_pos + 1);
+      bool our_file = (file_uuid == uuid_);
+
+      try {
+        long file_timestamp = std::stol(file_timestamp_str);
+        // We need to delete the file in 2 cases:
+        // 1. This is our file and it is older than this instance's
+        // history_keep_interval
+        // 2. This is someone's else file, but it is older than the highest
+        // possible history_keep_interval
+        if (our_file && (file_timestamp < oldest_timestamp)) {
```

Comment:
We compare time now (std::time(0)) (long) with the timestamp recovered from the filename (long), so it is the shortest path that avoids too much conversions.
We also create the file timestamp from std::time(0), which is long. In other way we would need to do
```
const auto now = std::chrono::system_clock::now();
const std::time_t t_c = std::chrono::system_clock::to_time_t(now);
```
which is actually what std::time(0) does

Reply:
Up to you but a few comments:
1. std::time_t is an unspecified type, so it may not be long.
2. Even if it is long, this is a SIGNED type which may not be what you expect (as you may have already noticed I don't like signed types :) )
3. For instance a difference between 2 timestamps may be negative (in case of clock change / ntp adjustments).